File: api/src/main/java/org/apache/iceberg/PartitionSpec.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iceberg;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Multimaps;
import com.google.common.collect.Sets;
import java.io.Serializable;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.apache.iceberg.exceptions.ValidationException;
import org.apache.iceberg.transforms.Transforms;
import org.apache.iceberg.transforms.UnknownTransform;
import org.apache.iceberg.types.Type;
import org.apache.iceberg.types.Types;
import org.apache.iceberg.types.Types.StructType;

/**
 * Represents how to produce partition data for a table.
 * <p>
 * Partition data is produced by transforming columns in a table. Each column transform is
 * represented by a named {@link PartitionField}.
 */
public class PartitionSpec implements Serializable {
  // start assigning IDs for partition fields at 1000
  private static final int PARTITION_DATA_ID_START = 1000;

  private final Schema schema;

  // this is ordered so that DataFile has a consistent schema
  private final int specId;
  private final PartitionField[] fields;
  private transient volatile ListMultimap<Integer, PartitionField> fieldsBySourceId = null;
  private transient volatile Map<String, PartitionField> fieldsByName = null;
  private transient volatile Class<?>[] lazyJavaClasses = null;
  private transient volatile List<PartitionField> fieldList = null;

  private PartitionSpec(Schema schema, int specId, List<PartitionField> fields) {
    this.schema = schema;
    this.specId = specId;
    this.fields = new PartitionField[fields.size()];
    for (int i = 0; i < this.fields.length; i += 1) {
      this.fields[i] = fields.get(i);
    }
  }

  /**
   * @return the {@link Schema} for this spec.
   */
  public Schema schema() {
    return schema;
  }

  /**
   * @return the ID of this spec
   */
  public int specId() {
    return specId;
  }

  /**
   * @return the list of {@link PartitionField partition fields} for this spec.
   */
  public List<PartitionField> fields() {
    return lazyFieldList();
  }

  /**
   * @param fieldId a field id from the source schema
   * @return the {@link PartitionField field} that partitions the given source field
   */
  public List<PartitionField> getFieldsBySourceId(int fieldId) {
    return lazyFieldsBySourceId().get(fieldId);
  }

  /**
   * @return a {@link StructType} for partition data defined by this spec.
   */
  public StructType partitionType() {
    List<Types.NestedField> structFields = Lists.newArrayListWithExpectedSize(fields.length);

    for (int i = 0; i < fields.length; i += 1) {
      PartitionField field = fields[i];
      Type sourceType = schema.findType(field.sourceId());
      Type resultType = field.transform().getResultType(sourceType);
      // assign ids for partition fields starting at PARTITION_DATA_ID_START to leave room for data file's other fields
      structFields.add(
          Types.NestedField.optional(PARTITION_DATA_ID_START + i, field.name(), resultType));
    }

    return Types.StructType.of(structFields);
  }

  public Class<?>[] javaClasses() {
    if (lazyJavaClasses == null) {
      synchronized (this) {
        if (lazyJavaClasses == null) {
          Class<?>[] classes = new Class<?>[fields.length];
          for (int i = 0; i < fields.length; i += 1) {
            PartitionField field = fields[i];
            if (field.transform() instanceof UnknownTransform) {
              classes[i] = Object.class;
            } else {
              Type sourceType = schema.findType(field.sourceId());
              Type result = field.transform().getResultType(sourceType);
              classes[i] = result.typeId().javaClass();
            }
          }

          this.lazyJavaClasses = classes;
        }
      }
    }

    return lazyJavaClasses;
  }

  @SuppressWarnings("unchecked")
  private <T> T get(StructLike data, int pos, Class<?> javaClass) {
    return data.get(pos, (Class<T>) javaClass);
  }

  private String escape(String string) {
    try {
      return URLEncoder.encode(string, "UTF-8");
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException(e);
    }
  }

  public String partitionToPath(StructLike data) {
    StringBuilder sb = new StringBuilder();
    Class<?>[] javaClasses = javaClasses();
    for (int i = 0; i < javaClasses.length; i += 1) {
      PartitionField field = fields[i];
      String valueString = field.transform().toHumanString(get(data, i, javaClasses[i]));

      if (i > 0) {
        sb.append("/");
      }
      sb.append(field.name()).append("=").append(escape(valueString));
    }
    return sb.toString();
  }

  /**
   * Returns true if this spec is equivalent to the other, with field names ignored. That is, if
   * both specs have the same number of fields, field order, source columns, and transforms.
   *
   * @param other another PartitionSpec
   * @return true if the specs have the same fields, source columns, and transforms.
   */
  public boolean compatibleWith(PartitionSpec other) {
    if (equals(other)) {
      return true;
    }

    if (fields.length != other.fields.length) {
      return false;
    }

    for (int i = 0; i < fields.length; i += 1) {
      PartitionField thisField = fields[i];
      PartitionField thatField = other.fields[i];
      if (thisField.sourceId() != thatField.sourceId() ||
          !thisField.transform().toString().equals(thatField.transform().toString())) {
        return false;
      }
    }

    return true;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (!(other instanceof PartitionSpec)) {
      return false;
    }

    PartitionSpec that = (PartitionSpec) other;
    if (this.specId != that.specId) {
      return false;
    }
    return Arrays.equals(fields, that.fields);
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(Arrays.hashCode(fields));
  }

  private List<PartitionField> lazyFieldList() {
    if (fieldList == null) {
      synchronized (this) {
        if (fieldList == null) {
          this.fieldList = ImmutableList.copyOf(fields);
        }
      }
    }
    return fieldList;
  }

  private Map<String, PartitionField> lazyFieldsByName() {
    if (fieldsByName == null) {
      synchronized (this) {
        if (fieldsByName == null) {
          ImmutableMap.Builder<String, PartitionField> builder = ImmutableMap.builder();
          for (PartitionField field : fields) {
            builder.put(field.name(), field);
          }
          this.fieldsByName = builder.build();
        }
      }
    }

    return fieldsByName;
  }

  private ListMultimap<Integer, PartitionField> lazyFieldsBySourceId() {
    if (fieldsBySourceId == null) {
      synchronized (this) {
        if (fieldsBySourceId == null) {
          ListMultimap<Integer, PartitionField> multiMap = Multimaps
              .newListMultimap(Maps.newHashMap(), () -> Lists.newArrayListWithCapacity(fields.length));
          for (PartitionField field : fields) {
            multiMap.put(field.sourceId(), field);
          }
          this.fieldsBySourceId = multiMap;
        }
      }
    }

    return fieldsBySourceId;
  }

  /**
   * Returns the source field ids for identity partitions.
   *
   * @return a set of source ids for the identity partitions.
   */
  public Set<Integer> identitySourceIds() {
    Set<Integer> sourceIds = Sets.newHashSet();
    for (PartitionField field : fields()) {
      if ("identity".equals(field.transform().toString())) {
        sourceIds.add(field.sourceId());
      }
    }

    return sourceIds;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("[");
    for (PartitionField field : fields) {
      sb.append("\n");
      sb.append("  ").append(field);
    }
    if (fields.length > 0) {
      sb.append("\n");
    }
    sb.append("]");
    return sb.toString();
  }

  private static final PartitionSpec UNPARTITIONED_SPEC =
      new PartitionSpec(new Schema(), 0, ImmutableList.of());

  /**
   * Returns a spec for unpartitioned tables.
   *
   * @return a partition spec with no partitions
   */
  public static PartitionSpec unpartitioned() {
    return UNPARTITIONED_SPEC;
  }

  /**
   * Creates a new {@link Builder partition spec builder} for the given {@link Schema}.
   *
   * @param schema a schema
   * @return a partition spec builder for the given schema
   */
  public static Builder builderFor(Schema schema) {
    return new Builder(schema);
  }

  /**
   * Used to create valid {@link PartitionSpec partition specs}.
   * <p>
   * Call {@link #builderFor(Schema)} to create a new builder.
   */
  public static class Builder {
    private final Schema schema;
    private final List<PartitionField> fields = Lists.newArrayList();
    private final Set<String> partitionNames = Sets.newHashSet();
    private Map<Integer, PartitionField> timeFields = Maps.newHashMap();
    private int specId = 0;

    private Builder(Schema schema) {
      this.schema = schema;
    }

    private void checkAndAddPartitionName(String name) {
      checkAndAddPartitionName(name, null);
    }

    private void checkAndAddPartitionName(String name, Integer identitySourceColumnId) {
      Types.NestedField schemaField = schema.findField(name);
      if (identitySourceColumnId != null) {
        // for identity transform case we allow  conflicts between partition and schema field name as
        //   long as they are sourced from the same schema field
        Preconditions.checkArgument(schemaField == null || schemaField.fieldId() == identitySourceColumnId,
            "Cannot create identity partition sourced from different field in schema: %s", name);
      } else {
        // for all other transforms we don't allow conflicts between partition name and schema field name
        Preconditions.checkArgument(schemaField == null,
            "Cannot create partition from name that exists in schema: %s", name);
      }
      Preconditions.checkArgument(name != null && !name.isEmpty(),
          "Cannot use empty or null partition name: %s", name);
      Preconditions.checkArgument(!partitionNames.contains(name),
          "Cannot use partition name more than once: %s", name);
      partitionNames.add(name);
    }

    private void checkForRedundantPartitions(PartitionField field) {
      PartitionField timeField = timeFields.get(field.sourceId());
      Preconditions.checkArgument(timeField == null,
          "Cannot add redundant partition: %s conflicts with %s", timeField, field);
      timeFields.put(field.sourceId(), field);
    }

    public Builder withSpecId(int newSpecId) {
      this.specId = newSpecId;
      return this;
    }

    private Types.NestedField findSourceColumn(String sourceName) {
      Types.NestedField sourceColumn = schema.findField(sourceName);
      Preconditions.checkArgument(sourceColumn != null, "Cannot find source column: %s", sourceName);
      return sourceColumn;
    }

    Builder identity(String sourceName, String targetName) {
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      checkAndAddPartitionName(targetName, sourceColumn.fieldId());
      fields.add(new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.identity(sourceColumn.type())));
      return this;
    }

    public Builder identity(String sourceName) {
      return identity(sourceName, sourceName);
    }

    public Builder year(String sourceName, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      PartitionField field = new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.year(sourceColumn.type()));
      checkForRedundantPartitions(field);
      fields.add(field);
      return this;
    }

    public Builder year(String sourceName) {
      return year(sourceName, sourceName + "_year");
    }

    public Builder month(String sourceName, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      PartitionField field = new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.month(sourceColumn.type()));
      checkForRedundantPartitions(field);
      fields.add(field);
      return this;
    }

    public Builder month(String sourceName) {
      return month(sourceName, sourceName + "_month");
    }

    public Builder day(String sourceName, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      PartitionField field = new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.day(sourceColumn.type()));
      checkForRedundantPartitions(field);
      fields.add(field);
      return this;
    }

    public Builder day(String sourceName) {
      return day(sourceName, sourceName + "_day");
    }

    public Builder hour(String sourceName, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      PartitionField field = new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.hour(sourceColumn.type()));
      checkForRedundantPartitions(field);
      fields.add(field);
      return this;
    }

    public Builder hour(String sourceName) {
      return hour(sourceName, sourceName + "_hour");
    }

    public Builder bucket(String sourceName, int numBuckets, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      fields.add(new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.bucket(sourceColumn.type(), numBuckets)));
      return this;
    }

    public Builder bucket(String sourceName, int numBuckets) {
      return bucket(sourceName, numBuckets, sourceName + "_bucket");
    }

    public Builder truncate(String sourceName, int width, String targetName) {
      checkAndAddPartitionName(targetName);
      Types.NestedField sourceColumn = findSourceColumn(sourceName);
      fields.add(new PartitionField(
          sourceColumn.fieldId(), targetName, Transforms.truncate(sourceColumn.type(), width)));
      return this;
    }

    public Builder truncate(String sourceName, int width) {
      return truncate(sourceName, width, sourceName + "_trunc");
    }

    Builder add(int sourceId, String name, String transform) {
      Types.NestedField column = schema.findField(sourceId);
      checkAndAddPartitionName(name, column.fieldId());
      Preconditions.checkNotNull(column, "Cannot find source column: %s", sourceId);
      fields.add(new PartitionField(sourceId, name, Transforms.fromString(column.type(), transform)));
      return this;
    }

    public PartitionSpec build() {
      PartitionSpec spec = new PartitionSpec(schema, specId, fields);
      checkCompatibility(spec, schema);
      return spec;
    }
  }

  static void checkCompatibility(PartitionSpec spec, Schema schema) {
    for (PartitionField field : spec.fields) {
      Type sourceType = schema.findType(field.sourceId());
      ValidationException.check(sourceType != null,
          "Cannot find source column for partition field: %s", field);
      ValidationException.check(sourceType.isPrimitiveType(),
          "Cannot partition by non-primitive source field: %s", sourceType);
      ValidationException.check(
          field.transform().canTransform(sourceType),
          "Invalid source type %s for transform: %s",
          sourceType, field.transform());
    }
  }
}

File: python/iceberg/api/partition_spec.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from urllib.parse import quote_plus

from iceberg.exceptions import ValidationException

from .partition_field import PartitionField
from .schema import Schema
from .transforms import Transform, Transforms
from .types import (NestedField,
                    StructType)


class PartitionSpec(object):

    PARTITION_DATA_ID_START = 1000

    @staticmethod
    def UNPARTITIONED_SPEC():
        return PartitionSpec(Schema(), 0, [], PartitionSpec.PARTITION_DATA_ID_START - 1)

    @staticmethod
    def unpartitioned():
        return PartitionSpec.UNPARTITIONED_SPEC()

    def __init__(self, schema, spec_id, fields, last_assigned_field_id):
        self.fields_by_source_id = None
        self.fields_by_name = None
        self.__java_classes = None
        self.field_list = None

        self.schema = schema
        self.spec_id = spec_id
        self.__fields = list()
        for field in fields:
            self.__fields.append(field)
        self.last_assigned_field_id = last_assigned_field_id

    @property
    def fields(self):
        return self.lazy_field_list()

    @property
    def java_classes(self):
        if self.__java_classes is None:
            self.__java_classes
        for field in self.__fields:
            source_type = self.schema.find_type(field.source_id)
            result = field.transform().get_result_by_type(source_type)
            self.__java_classes.append(result.type_id.java_class())

        return self.__java_classes

    def get_field_by_source_id(self, field_id):
        return self.lazy_fields_by_source_id().get(field_id)

    def partition_type(self):
        struct_fields = list()
        for _i, field in enumerate(self.__fields):
            source_type = self.schema.find_type(field.source_id)
            result_type = field.transform.get_result_type(source_type)
            struct_fields.append(NestedField.optional(field.field_id,
                                                      field.name,
                                                      result_type))

        return StructType.of(struct_fields)

    def get(self, data, pos, java_class):
        data.get(pos, java_class)

    def escape(self, string):
        return quote_plus(string, encoding="UTF-8")

    def partition_to_path(self, data):
        sb = list()
        java_classes = self.java_classes
        for i, jclass in enumerate(java_classes):
            field = self.__fields[i]
            value_string = field.transform().to_human_string(self.get(data, i, jclass))

            if i > 0:
                sb.append("/")
            sb.append(field.name)
            sb.append("=")
            sb.append(self.escape(value_string))

        return "".join(sb)

    def compatible_with(self, other):
        if self.__eq__(other):
            return True

        if len(self.__fields) != len(other.__fields):
            return False

        for i, field in enumerate(self.__fields):
            that_field = other.__fields[i]
            if field.source_id != that_field.source_id or str(field.transform) != str(that_field.transform):
                return False

        return True

    def lazy_fields_by_source_id(self):
        if self.fields_by_source_id is None:
            self.fields_by_source_id = dict()
            for field in self.fields:
                self.fields_by_source_id[field.source_id] = field

        return self.fields_by_source_id

    def identity_source_ids(self):
        source_ids = set()
        fields = self.fields
        for field in fields:
            if "identity" == str(field.transform()):
                source_ids.add(field)

        return source_ids

    def lazy_field_list(self):
        if self.field_list is None:
            self.field_list = list(self.__fields)

        return self.field_list

    def lazy_fields_by_source_name(self):
        if self.fields_by_name is None:
            self.fields_by_name = dict()
            for field in self.__fields:
                self.fields_by_name[field.name] = field

        return self.fields_by_name

    def __eq__(self, other):
        if id(self) == id(other):
            return True

        if other is None or not isinstance(other, PartitionSpec):
            return False

        return self.__fields == other.__fields

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return hash(self.__key())

    def __key(self):
        return PartitionSpec.__class__, tuple(self.fields)

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        sb = ["["]

        for field in self.__fields:
            sb.append("\n {field_id}: {name}: {transform}({source_id})".format(field_id=field.field_id,
                                                                               name=field.name,
                                                                               transform=str(field.transform),
                                                                               source_id=field.source_id))

        if len(self.__fields) > 0:
            sb.append("\n")
        sb.append("]")

        return "".join(sb)

    @staticmethod
    def builder_for(schema: Schema) -> "PartitionSpecBuilder":
        return PartitionSpecBuilder(schema)

    @staticmethod
    def check_compatibility(spec, schema):
        for field in spec.fields:
            src_type = schema.find_type(field.source_id)
            if not src_type.is_primitive_type():
                raise ValidationException("Cannot partition by non-primitive source field: %s", src_type)
            if not field.transform.can_transform(src_type):
                raise ValidationException("Invalid source type %s for transform: %s", (src_type, field.transform))


class PartitionSpecBuilder(object):

    def __init__(self, schema):
        self.schema = schema
        self.fields = list()
        self.partition_names = set()
        self.dedup_fields = dict()
        self.spec_id = 0
        self.last_assigned_field_id = PartitionSpec.PARTITION_DATA_ID_START - 1

    def __next_field_id(self):
        self.last_assigned_field_id = self.last_assigned_field_id + 1
        return self.last_assigned_field_id

    def with_spec_id(self, spec_id):
        self.spec_id = spec_id
        return self

    def check_and_add_partition_name(self, name, source_column_id=None):
        schema_field = self.schema.find_field(name)
        if source_column_id is not None:
            if schema_field is not None and schema_field.field_id != source_column_id:
                raise ValueError("Cannot create identity partition sourced from different field in schema: %s" % name)
        else:
            if schema_field is not None:
                raise ValueError("Cannot create partition from name that exists in schema: %s" % name)

        if name is None or name == "":
            raise ValueError("Cannot use empty or null partition name: %s" % name)
        if name in self.partition_names:
            raise ValueError("Cannot use partition names more than once: %s" % name)

        self.partition_names.add(name)
        return self

    def check_redundant_and_add_field(self, field_id: int, name: str, transform: Transform) -> None:
        field = PartitionField(field_id,
                               self.__next_field_id(),
                               name,
                               transform)
        dedup_key = (field.source_id, field.transform.dedup_name())
        partition_field = self.dedup_fields.get(dedup_key)
        if partition_field is not None:
            raise ValueError("Cannot add redundant partition: %s conflicts with %s" % (partition_field, field))
        self.dedup_fields[dedup_key] = field
        self.fields.append(field)

    def find_source_column(self, source_name):
        source_column = self.schema.find_field(source_name)
        if source_column is None:
            raise RuntimeError("Cannot find source column: %s" % source_name)

        return source_column

    def identity(self, source_name, target_name=None):
        if target_name is None:
            target_name = source_name

        source_column = self.find_source_column(source_name)
        self.check_and_add_partition_name(target_name, source_column.field_id)
        self.check_redundant_and_add_field(source_column.field_id,
                                           target_name,
                                           Transforms.identity(source_column.type))
        return self

    def year(self, source_name, target_name=None):
        if target_name is None:
            target_name = "{}_year".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.check_redundant_and_add_field(source_column.field_id,
                                           target_name,
                                           Transforms.year(source_column.type))
        return self

    def month(self, source_name, target_name=None):
        if target_name is None:
            target_name = "{}_month".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.check_redundant_and_add_field(source_column.field_id,
                                           target_name,
                                           Transforms.month(source_column.type))
        return self

    def day(self, source_name, target_name=None):
        if target_name is None:
            target_name = "{}_day".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.check_redundant_and_add_field(source_column.field_id,
                                           target_name,
                                           Transforms.day(source_column.type))
        return self

    def hour(self, source_name, target_name=None):
        if target_name is None:
            target_name = "{}_hour".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.check_redundant_and_add_field(source_column.field_id,
                                           target_name,
                                           Transforms.hour(source_column.type))
        return self

    def bucket(self, source_name, num_buckets, target_name=None):
        if target_name is None:
            target_name = "{}_bucket".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.fields.append(PartitionField(source_column.field_id,
                                          self.__next_field_id(),
                                          target_name,
                                          Transforms.bucket(source_column.type, num_buckets)))
        return self

    def truncate(self, source_name, width, target_name=None):
        if target_name is None:
            target_name = "{}_truncate".format(source_name)

        self.check_and_add_partition_name(target_name)
        source_column = self.find_source_column(source_name)
        self.fields.append(PartitionField(source_column.field_id,
                                          self.__next_field_id(),
                                          target_name,
                                          Transforms.truncate(source_column.type, width)))
        return self

    def add_without_field_id(self, source_id, name, transform):
        return self.add(source_id, self.__next_field_id(), name, transform)

    def add(self, source_id: int, field_id: int, name: str, transform: str) -> "PartitionSpecBuilder":
        column = self.schema.find_field(source_id)
        if column is None:
            raise ValueError("Cannot find source column: %s" % source_id)

        transform_obj = Transforms.from_string(column.type, transform)
        self.check_and_add_partition_name(name, source_id)
        self.fields.append(PartitionField(source_id,
                                          field_id,
                                          name,
                                          transform_obj))
        self.last_assigned_field_id = max(self.last_assigned_field_id, field_id)
        return self

    def build(self):
        spec = PartitionSpec(self.schema, self.spec_id, self.fields, self.last_assigned_field_id)
        PartitionSpec.check_compatibility(spec, self.schema)
        return spec
